### PR TITLE
Add basic utilities for new implementation of justification heuristic

### PR DESCRIPTION
**Pull request by ajreynol**
This sets up the core utilities for the new implementation of justification heuristic

### CODE REVIEW

File: src/decision/justify_info.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Information concerning a single formula in the justification strategy.
+ */
+class JustifyInfo
+{
+ public:
+  JustifyInfo(context::Context* c);
+  ~JustifyInfo();
+  /** set */
+  void set(TNode n, prop::SatValue desiredVal);
```
Comment (mpreiner) [inline src/decision/justify_info.h:39]: Can we make `n` const?

File: src/decision/justify_stack.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  /** The current assertion we are trying to satisfy */
+  context::CDO<TNode> d_current;
+  /**
+   * Stack of justify info, valid up to index d_stackSizeValid-1. Notice the
+   * size of this list may be larger than the current size we are using in
+   * cases where we considered an assertion requiring a larger stack size
+   * than the current one. This is because we do not erase elements from
+   * CDList in a context-dependent manner.
+   */
+  context::CDList<std::shared_ptr<JustifyInfo> > d_stack;
```
Comment (mpreiner) [inline src/decision/justify_stack.h:72]: Why a shared ptr and not a unique ptr?

**Comment by mpreiner** [comment]
What I forgot to add: I suggest to add this to a `decision` namespace.

File: src/decision/justify_stack.h (hunk 2 above)
Comment (ajreynol) [reply src/decision/justify_stack.h:72]: This is a current limitation of `context::CDList` unfortunately.  I'm not sure the details of why this is not possible, maybe @4tXJ7f can say?

File: src/decision/justify_info.h (hunk 1 above)
Comment (ajreynol) [reply src/decision/justify_info.h:39]: I think the option would be to change this to `const Node& n` right?

**Review by 4tXJ7f — APPROVED** [review]
LGTM, other than the discussions.

**Comment by ajreynol** [comment]
> LGTM, other than the discussions.

**Comment by ajreynol** [comment]
@4tXJ7f `CDList<JustifyInfo>` leads to several compilation errors:
```
In file included from /usr/include/c++/10/x86_64-redhat-linux/bits/c++allocator.h:33,
                 from /usr/include/c++/10/bits/allocator.h:46,
                 from /usr/include/c++/10/string:41,
                 from /usr/include/c++/10/bits/locale_classes.h:40,
                 from /usr/include/c++/10/bits/ios_base.h:41,
                 from /usr/include/c++/10/streambuf:41,
                 from /usr/include/c++/10/bits/streambuf_iterator.h:35,
                 from /usr/include/c++/10/iterator:66,
                 from /space/ajreynol/cvc4-pr-ajr/src/./context/cdlist.h:24,
                 from /space/ajreynol/cvc4-pr-ajr/src/./decision/justify_stack.h:20,
                 from /space/ajreynol/cvc4-pr-ajr/src/decision/justify_stack.cpp:15:
/usr/include/c++/10/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = cvc5::decision::JustifyInfo; _Args = {const cvc5::decision::JustifyInfo&}; _Tp = cvc5::decision::JustifyInfo]’:
/space/ajreynol/cvc4-pr-ajr/src/./context/cdlist.h:294:26:   required from ‘void cvc5::context::CDList<T, CleanUp, Allocator>::push_back(const T&) [with T = cvc5::decision::JustifyInfo; CleanUpT = cvc5::context::DefaultCleanUp<cvc5::decision::JustifyInfo>; AllocatorT = std::allocator<cvc5::decision::JustifyInfo>]’
/space/ajreynol/cvc4-pr-ajr/src/decision/justify_stack.cpp:83:45:   required from here
/usr/include/c++/10/ext/new_allocator.h:150:4: error: use of deleted function ‘cvc5::decision::JustifyInfo::JustifyInfo(const cvc5::decision::JustifyInfo&)’
  150 |  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /space/ajreynol/cvc4-pr-ajr/src/./decision/justify_stack.h:22,
                 from /space/ajreynol/cvc4-pr-ajr/src/decision/justify_stack.cpp:15:
/space/ajreynol/cvc4-pr-ajr/src/./decision/justify_info.h:33:7: note: ‘cvc5::decision::JustifyInfo::JustifyInfo(const cvc5::decision::JustifyInfo&)’ is implicitly deleted because the default definition would be ill-formed:
   33 | class JustifyInfo
      |       ^~~~~~~~~~~
/space/ajreynol/cvc4-pr-ajr/src/./decision/justify_info.h:33:7: error: ‘cvc5::context::CDO<T>::CDO(const cvc5::context::CDO<T>&) [with T = cvc5::NodeTemplate<false>]’ is protected within this context
In file included from /space/ajreynol/cvc4-pr-ajr/src/./decision/justify_stack.h:21,
                 from /space/ajreynol/cvc4-pr-ajr/src/decision/justify_stack.cpp:15:
/space/ajreynol/cvc4-pr-ajr/src/./context/cdo.h:45:3: note: declared protected here
   45 |   CDO(const CDO<T>& cdo) : ContextObj(cdo), d_data(cdo.d_data) {}
      |   ^~~
In file included from /space/ajreynol/cvc4-pr-ajr/src/./decision/justify_stack.h:22,
                 from /space/ajreynol/cvc4-pr-ajr/src/decision/justify_stack.cpp:15:
/space/ajreynol/cvc4-pr-ajr/src/./decision/justify_info.h:33:7: error: ‘cvc5::context::CDO<T>::CDO(const cvc5::context::CDO<T>&) [with T = cvc5::prop::SatValue]’ is protected within this context
   33 | class JustifyInfo
      |       ^~~~~~~~~~~
In file included from /space/ajreynol/cvc4-pr-ajr/src/./decision/justify_stack.h:21,
                 from /space/ajreynol/cvc4-pr-ajr/src/decision/justify_stack.cpp:15:
/space/ajreynol/cvc4-pr-ajr/src/./context/cdo.h:45:3: note: declared protected here
   45 |   CDO(const CDO<T>& cdo) : ContextObj(cdo), d_data(cdo.d_data) {}
      |   ^~~
In file included from /space/ajreynol/cvc4-pr-ajr/src/./decision/justify_stack.h:22,
                 from /space/ajreynol/cvc4-pr-ajr/src/decision/justify_stack.cpp:15:
/space/ajreynol/cvc4-pr-ajr/src/./decision/justify_info.h:33:7: error: ‘cvc5::context::CDO<T>::CDO(const cvc5::context::CDO<T>&) [with T = long unsigned int]’ is protected within this context
   33 | class JustifyInfo
      |       ^~~~~~~~~~~
In file included from /space/ajreynol/cvc4-pr-ajr/src/./decision/justify_stack.h:21,
                 from /space/ajreynol/cvc4-pr-ajr/src/decision/justify_stack.cpp:15:
/space/ajreynol/cvc4-pr-ajr/src/./context/cdo.h:45:3: note: declared protected here
   45 |   CDO(const CDO<T>& cdo) : ContextObj(cdo), d_data(cdo.d_data) {}
      |   ^~~
```

**Comment by 4tXJ7f** [comment]
Ah, it looks like `CDO` is not copy-able (makes sense). With my PR, it may be possible using `CDList::emplace_back()` since that avoids the issue of copying `CDO` instances.

**Comment by ajreynol** [comment]
Cool, I'll update this after your PR is in.

**Pull request merged**